kotlin/src/com/here/flexiblepolyline/FlexiblePolylineTest.kt

```diff
@@ -0,0 +1,380 @@
+            test.testSimpleLatLngEncoding()
+            test.testComplexLatLngEncoding()
+            test.testLatLngZEncode()
+            test.encodingSmokeTest()
+
+            //Decode test
+            test.testInvalidEncoderInput()
+            test.testThirdDimension()
+            test.testDecodeConvertValue()
+            test.testSimpleLatLngDecoding()
+            test.testComplexLatLngDecoding()
+            test.testLatLngZDecode()
+            test.decodingSmokeTest()
+            test.testVeryLongLine(lineLength)
+        }
+    }
+}
+
+private data class TestData(
+    val precision: Int = 0,
+    val thirdDimensionPrecision: Int = 0,
+    val thirdDimension: FlexiblePolyline.ThirdDimension? = null,
+    val latLngZs: List<FlexiblePolyline.LatLngZ>? = null
+){}
+
+private data class TestCase(
+    val testInput: String,
+    val testResult: String
+){}
+
+private class TestCaseReader(testInputFile: String, testResultFile: String) : Iterator<TestCase> {
+    private var totalLines = 0
+    private var currentLine = 0
+    private var testCases = mutableListOf<TestCase>()
+
+    init {
+        try {
+            Files.newBufferedReader(Paths.get(FlexiblePolylineTest.TEST_FILES_RELATIVE_PATH + testInputFile)).use { input ->
+                Files.newBufferedReader(Paths.get(FlexiblePolylineTest.TEST_FILES_RELATIVE_PATH + testResultFile)).use { result ->
+                    // read line by line and validate the test
+                    while (true) {
+                        val regex = "\\s|\\(|\\)".toRegex()
+                        val testInputFileLine = input.readLine();
+                        val testResultFileLine = result.readLine();
+
+                        if (testInputFileLine != null && testInputFileLine.isNotBlank() && testResultFileLine != null && testResultFileLine.isNotBlank()) {
+                            testCases.add(
+                                TestCase(
+                                    testInput = testInputFileLine.replace(regex, ""),
+                                    testResult = testResultFileLine.replace(regex, "")
+                                )
+                            )
+                            totalLines++
+                        } else {
+                            break
+                        }
+                    }
+                }
+            }
+        } catch (e: Exception) {
+            e.printStackTrace()
+            System.err.format("TestCaseReader - exception reading test case $testInputFile and $testResultFile at LineNo: $totalLines: %s%n", e)
+            throw RuntimeException("Test failed, as test data could not be loaded by TestCaseReader")
+        }
+    }
+
+    override fun hasNext(): Boolean {
+        return currentLine < totalLines
+    }
+
+    override fun next(): TestCase {
+        if(!hasNext()) throw NoSuchElementException()
+        val testCase = testCases[currentLine]
+        currentLine++
+        return testCase
+    }
```

Comment:
Instead of making the whole class an iterator, I would suggest a method: getTestCases() which returns a List<TestCase> or Iterator<TestCase>. In any case I would reuse existing iterator implementation for List from the standard library.

Reply:
Done.